nautobot/core/api/serializers.py

```diff
@@ -132,14 +171,82 @@ def get_field_names(self, declared_fields, info):
+            fields.append("object_type")
         return fields
 
+    def build_field(self, field_name, info, model_class, nested_depth):
+        """
+        Return a two tuple of (cls, kwargs) to build a serializer field with.
+        """
+        from rest_framework.utils.model_meta import RelationInfo, _get_to_field
+
+        # For tags field, DRF does not recognize the relationship between tags and the model itself (?)
+        # so instead of calling build_nested_field() it will call build_property_field() which
+        # makes the field impervious to the `?depth` parameter.
+        # So we intercept it here to call build_nested_field()
+        # which will make the tags field be rendered with TagSerializer() and respect the `depth` parameter.
+        if field_name == "tags":
+            if nested_depth > 0:
+                relation_info = RelationInfo(
+                    model_field=getattr(model_class, "tags"),
+                    related_model=Tag,
+                    to_many=True,
+                    has_through_model=True,
+                    to_field=_get_to_field(getattr(model_class, "tags")),
+                    reverse=False,
+                )
+                return self.build_nested_field(field_name, relation_info, nested_depth)
+
+        return super().build_field(field_name, info, model_class, nested_depth)
+
+    def build_property_field(self, field_name, model_class):
+        """
+        Create a property field for model methods and properties.
+        """
+        if field_name == "tags":
+            field_class = NautobotPrimaryKeyRelatedField
+            field_kwargs = {
+                "queryset": Tag.objects.filter(content_types=ContentType.objects.get_for_model(model_class)),
+                "required": False,
+            }
+
+            return field_class, field_kwargs
+        return super().build_property_field(field_name, model_class)
+
+    def build_nested_field(self, field_name, relation_info, nested_depth):
+        field = get_serializer_for_model(relation_info.related_model)
+
+        class NautobotNestedSerializer(field):
+            class Meta:
+                model = relation_info.related_model
+                depth = nested_depth - 1
+                if hasattr(field.Meta, "fields"):
+                    fields = field.Meta.fields
+                if hasattr(field.Meta, "exclude"):
+                    exclude = field.Meta.exclude
+
+        # This is a very hacky way to avoid name collisions in OpenAPISchema Generations
+        # The exact error output can be seen in this issue https://github.com/tfranzel/drf-spectacular/issues/90
+        # Apparently drf-spectacular does not support the `?depth` argument that comes with DRF
+        # So auto-generating NestedSerializers with the default class names that are the same when depth > 0
+        # does not make our schema happy.
+        NautobotNestedSerializer.__name__ = "NautobotNestedSerializer" + f"{uuid.uuid1()}"
```

Comment:
> We have `field_name` and `relation_info.related_model` both available to us, we should be able to come up with a better name here. If the issue here is to avoid collisions between e.g. "Device.location" and "Cluster.location" both creating a `NestedLocationSerializer`, then the correct fix should be to maintain a registry of nested serializers we've already created and reuse them where appropriate, because they _should_ be the same serializer class - take a look at how `get_writable_class()` in `nautobot.core.api.schema` is implemented.

I do not know if a combination of `related_model` name and `model` name is sufficient to create a unique enough combination. Take `tags` on `location` for example. The list of tags represented by the same name `LocationTagsNautobotNestedSerializers` will still cause a conflict in `invoke check-schema`